Python/immutability.c

```diff
@@ -393,66 +368,89 @@ PyObject* _PyImmutability_Freeze(PyObject* obj)
             continue;
         }
 
-        if(!can_freeze(item->ob_type, state->freezable_types)){
-            PyObject* error_msg = PyUnicode_FromFormat("Cannot freeze object of type %s", Py_TYPE(item)->tp_name);
-            PyErr_SetObject(PyExc_TypeError, error_msg);
-            result = NULL;
-            goto cleanup;
+        if(PyObject_IsInstance(item, (PyObject*)&_PyNotFreezable_Type)){
+            PyErr_SetString(PyExc_TypeError, "Cannot freeze NotFreezable type");
+            goto error;
         }
 
         if(_Py_IsImmutable(item)){
             continue;
         }
 
-        _Py_SetImmutable(item);
-
         if(is_c_wrapper(item)) {
             // C functions are not mutable
             // Types are manually traversed
+            _Py_SetImmutable(item);
             continue;
         }
 
         if(PyFunction_Check(item)){
             if(shadow_function_globals(item) == NULL){
-                goto cleanup;
+                goto error;
             }
         }
 
         if(PyType_Check(item)){
             PyTypeObject* type = (PyTypeObject*)item;
+
             if(push(frontier, type->tp_dict))
             {
-                result = PyErr_NoMemory();
-                goto cleanup;
+                goto error;
             }
 
-            if(PySet_Contains(state->freezable_types, item) != 1){
-                // type is not explicit freezable, so we need to check its bases
-                if(push(frontier, type->tp_mro))
-                {
-                    result = PyErr_NoMemory();
-                    goto cleanup;
-                }
+            if(_Py_IsImmutable(item)){
+                continue;
+            }
+
+            _Py_SetImmutable(item);
+
+            if(is_freezable_builtin(type)){
+                _Py_SetImmutable(item);
+                continue;
+            }
+
+            if (PySet_Contains(state->freezable_types, item) == 1) {
+                _Py_SetImmutable(item);
+                continue;
+            }
```

Comment:
Could a freezable_type have non-trivial state hanging off it, and thus this would prevent that state from being frozen?  Should we still call tp_traverse on this to get that state.

Also, you seem to be calling SetImmutable a lot of times.